[ONNX] Create decorator to handle symbolic context

justinchuby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85488
* #85487
* #85486
* #85350
* #85349
* #85312
* #85165
* #85491
* __->__ #84776
* #84728

- Create decorator to handle old style custom symbolics that require context
- Deprecate `torch.onnx.SymbolicContext` in favor of `GraphContext`. Added deprecation message
- Remove README reference of SymbolicContext

justinchuby committed:
[ONNX] Create decorator to handle old style custom symbolics that require context

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle old style custom symbolics that require context"




[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle old style custom symbolics that require context"




[ghstack-poisoned]

AllenTiTaiWang commented on torch/onnx/_exporter_states.py:
```diff
@@ -25,6 +26,12 @@ class SymbolicContext:
+        "1.13",
+        "TBD",
+        # TODO(justinchuby): Fix the instruction when GraphContext is public.
+        "remove the 'ctx' argument and annotate 'g: GraphContext' instead",
```
I thought this was still under discusstion😂

AllenTiTaiWang commented on torch/onnx/_exporter_states.py:
```diff
@@ -25,6 +26,12 @@ class SymbolicContext:
+        "TBD",
+        # TODO(justinchuby): Fix the instruction when GraphContext is public.
+        "remove the 'ctx' argument and annotate 'g: GraphContext' instead",
+    )
```
Will `SymbolicContext` be deprecated?

justinchuby replied:
I intended to deprecate SymbolicContext. @BowenBao @shubhambhokare1 @thiagocrepaldi any thoughts?

BowenBao replied:
Sounds good, since all previous functionalities can be achieved through new GraphContext.

titaiwangms reviewed (approved):
LGTM!

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`

TODO: Unit test


[ghstack-poisoned]

thiagocrepaldi reviewed (commented):
@BowenBao and I were discussing a similar issue (passing context to some symbolics, but in a deeper level  - that we can discuss separately to prevent noise here).

In short, this context would *also* be populated by the JIT [script?] C++ core, with `dtype`s, `shape` and whatever is available so that it can assist symbolic implementors during `torch.onnx.export` for models with `torch.jit.script` layers. For scripted layers, hard-to-guess construct happen, such as `Tensor[]` in which we cannot know whether it was dtype=float or a list of something else, for example (that seems to be avaiable during torch.jit.script time).

Regarding to this PR in specific, I wonder if we could *always* make the existing context available to all symbolics by default, making it a generic solution for any new symbolic/opset to come. 

For this approach, an idea would be tapping the context into [Graph](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir/ir.h#L1165) and/or adding a [python binding method to Graph](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/python/python_ir.cpp#L218)

Although this is a change to the JIT IR core, if we do it in a nice way (low coupling and high cohesion), meta might accept changes to `tyorch._C.Graph`. 

With this, we might reuse it not only to solve the current problem (some symbolics need params_dict, env, cur_node, onnx_block), but also will have a standard way to extend `torch.jit.script` support, by providing means to the core to push context there too

@BowenBao what do you think?

justinchuby commented:
> Regarding to this PR in specific, I wonder if we could *always* make the existing context available to all symbolics by default, making it a generic solution for any new symbolic/opset to come. 

> tapping the context into Graph 

Great idea! Please see #84728 (previous PR on the stack) that implements this. Can you check the discussion section in the PR and share what you think? The decorator exists to support the legacy style (for bc) that specifies an additional context. We expect all new symbolic functions assume the context in `g`.

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

thiagocrepaldi commented:
> > Regarding to this PR in specific, I wonder if we could _always_ make the existing context available to all symbolics by default, making it a generic solution for any new symbolic/opset to come.
> 
> > tapping the context into Graph
> 
> Great idea! Please see #84728 (previous PR on the stack) that implements this. Can you check the discussion section in the PR and share what you think? The decorator exists to support the legacy style (for bc) that specifies an additional context. We expect all new symbolic functions assume the context in `g`.

Add a deprecation message to the decorator to let users know about the new mechanism. This way we can try to remove the decorator in a couple releases.

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

BowenBao reviewed (approved):
LGTM, please add a few test cases for coverage.

BowenBao commented on torch/onnx/_exporter_states.py:
```diff
@@ -25,6 +26,12 @@ class SymbolicContext:
 
+    @_deprecation.deprecated(
+        "1.13",
+        "TBD",
```
Forgot to mention, do you intend to keep TBD?

justinchuby replied:
Updated to 1.14

justinchuby replied:
README updated. Public user doc to be updated in another PR.

BowenBao reviewed (changes requested):
Just noticed that SymbolicContext is also mentioned in our onnx doc page, please update the information there.

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby commented:
@BowenBao Do you think this test is good enough?

https://github.com/pytorch/pytorch/blob/848437590f41af5d3c9f9bb381106114f70fe572/test/onnx/test_custom_ops.py#L111-L129

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


Create decorator to handle old style custom symbolics that require context

TODO: Unit test


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


- Create decorator to handle old style custom symbolics that require context
- Deprecate `torch.onnx.SymbolicContext` in favor of `GraphContext`. Added deprecation message
- Remove README reference of SymbolicContext


[ghstack-poisoned]

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -45,6 +45,7 @@
     symbolic_helper,
 )
 from torch.onnx._globals import GLOBALS
+from torch.onnx import _deprecation
```
Remove

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1776,12 +1802,6 @@ def _run_symbolic_function(
-                    return symbolic_fn(ctx, graph_context, *inputs, **attrs)
                 # PythonOp symbolic need access to the node to resolve the name conflict,
                 # this is inconsistent with regular op symbolic.
                 if op_name == "PythonOp":
```
What happens if a PythonOp needs symbolic_context? @BowenBao how can we handle python ops elegantly?

justinchuby replied:
We can remove this.

justinchuby replied:
Removed

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


- Create decorator to handle old style custom symbolics that require context
- Deprecate `torch.onnx.SymbolicContext` in favor of `GraphContext`. Added deprecation message
- Remove README reference of SymbolicContext


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


- Create decorator to handle old style custom symbolics that require context
- Deprecate `torch.onnx.SymbolicContext` in favor of `GraphContext`. Added deprecation message
- Remove README reference of SymbolicContext


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


- Create decorator to handle old style custom symbolics that require context
- Deprecate `torch.onnx.SymbolicContext` in favor of `GraphContext`. Added deprecation message
- Remove README reference of SymbolicContext


[ghstack-poisoned]

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1707,6 +1708,32 @@ def _need_symbolic_context(symbolic_fn) -> bool:
+        # TODO(justinchuby): Update the module name of GraphContext when it is public
+        warnings.warn(
+            "The first argument to symbolic functions is deprecated in 1.13 and will be "
+            "removed in 1.14. Please annotate the first argument as GraphContext "
```
change wording

justinchuby replied:
Done

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


- Create decorator to handle old style custom symbolics that require context
- Deprecate `torch.onnx.SymbolicContext` in favor of `GraphContext`. Added deprecation message
- Remove README reference of SymbolicContext


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


- Create decorator to handle old style custom symbolics that require context
- Deprecate `torch.onnx.SymbolicContext` in favor of `GraphContext`. Added deprecation message
- Remove README reference of SymbolicContext


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


- Create decorator to handle old style custom symbolics that require context
- Deprecate `torch.onnx.SymbolicContext` in favor of `GraphContext`. Added deprecation message
- Remove README reference of SymbolicContext


[ghstack-poisoned]

justinchuby commented:
@BowenBao PTAL

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


- Create decorator to handle old style custom symbolics that require context
- Deprecate `torch.onnx.SymbolicContext` in favor of `GraphContext`. Added deprecation message
- Remove README reference of SymbolicContext


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


- Create decorator to handle old style custom symbolics that require context
- Deprecate `torch.onnx.SymbolicContext` in favor of `GraphContext`. Added deprecation message
- Remove README reference of SymbolicContext


[ghstack-poisoned]

BowenBao reviewed (approved):
🚀

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


- Create decorator to handle old style custom symbolics that require context
- Deprecate `torch.onnx.SymbolicContext` in favor of `GraphContext`. Added deprecation message
- Remove README reference of SymbolicContext


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create decorator to handle symbolic context"


- Create decorator to handle old style custom symbolics that require context
- Deprecate `torch.onnx.SymbolicContext` in favor of `GraphContext`. Added deprecation message
- Remove README reference of SymbolicContext


[ghstack-poisoned]

justinchuby commented:
@pytorchbot merge
